### DISCUSSION TITLE
Question About Best Practices With Relation Queries

### DISCUSSION THREAD

**Discussion by arobinsongit** [prisma/prisma#13653] [Q&A] [answered]
Working with a team of devs building out a greenfield application using Cockroach, NextJS (Vercel) and likely Prisma.  I'm working on the CRDB side and I have a question about best practices when using Relation Queries.

A very watered-down description of what I'm doing could be described as a content management platform.   So you have documents, versions of documents, sections, and content all in a relational DB.  My old school SQL Server brain says I should build a series of views, getting progressively wider and wider that simplify things for my front end team.   A couple examples connected examples below

```
CREATE OR REPLACE view
	content.vw_document
AS 
SELECT 
	d.id as documentid,dt.id as documenttypeid, d."name" as documentname, d.description as documentdescription, d.ownedby as documentowner,
	d.name AS localID,
	dt."name" as documenttype,
	c.name as containername, c.description  as containerdescription
from
	content.document d
inner join 
	content.documenttype dt on dt.id = d.typeid
inner join 
	content.container c on (c.id = d.parentid and d.parenttype = 'container')
;	

CREATE OR REPLACE view
	content.vw_documentversion
AS 
SELECT 
	dv.id as documentversionid, dv.version as documentversion, dv."comment" as versioncomment,  
	vd.documentid,vd.documenttypeid, vd."documentname", vd.documentdescription, vd.documentowner,
	vd.localid as localid,
	vd.documenttype as documenttype,
	vd.containername as containername, vd.containerdescription as containerdescription
from
	content.vw_document vd
inner join 
	content.documentversion dv on (vd.documentid = dv.documentid)
;
```

Pretty basic, just chaining a couple views together.  The most complex I get right now is using some windows functions to support outline style numbering

```
row_number() OVER (partition by vdv.documentversionid order by ds.sortorder ASC) as localID,
concat(vdv.localID,'.',cast(row_number() OVER (partition by ds.id) as string)) as combinedSectionID	
```

Here is the question.  If I correctly describe my joins using Prisma Client (https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries) under the covers is it really just writing the same SQL I am?  There are probably some subtle style differences but the plan ends up being basically the same?

If I go write my own views in CRDB and have Prisma hit those instead of using the client and describing the queries there am I really just throwing away all the benefits of Prisma as a tool to decouple my DB from my application?  I think the answer is yes but just looking for some confirmation.

Regards,
Andy

**Comment by austincrim** [comment, marked as answer]
Hey @arobinsongit 👋 ,

Prisma does not have first-class support for views ([existing issue](https://github.com/prisma/prisma/issues/678)), so you would need to write your view definitions in SQL no matter what.

[This doc](https://www.prisma.io/docs/guides/database/advanced-database-tasks/sql-views-postgres) walks through creating a view and manually adding a corresponding `model` for it so that it can be queried through the Prisma Client. This way you can still get autocomplete and type-safety when querying the view.

Let me know if I can clarify!

**Comment by arobinsongit** [comment]
Thanks for the clear answer @austincrim.  A definitive "no but here's another way to look at it" is just as valuable as a "yes but'.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
